Onboarding: Intermediate lectures (+ runnable Aiken and Mesh examples) (#1951)
cardano-foundation/developer-portal: docs/developers/onboarding/lectures/intermediate/introduction.md

Author: I am using it only in lists as the curriculum is using. What do you recommend instead?

---
title: "Intermediate: smart contracts"
sidebar_label: "Introduction"
description: "Smart contracts from scratch — on-chain vs off-chain, validators, datum and redeemer, the tools to write and run them, then vesting, gift cards, oracles and testing."
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Intermediate: smart contracts

You finished Beginner, so you can move value around Cardano. This track makes the chain **enforce rules** about how that value moves. That is what a smart contract is. We build up from the ideas to a real contract you compile and run yourself.

:::note Coming from Ethereum?

> "Smart contract" means something different here. On Cardano it is not a deployed program with storage that you call and that then acts. It is a **rule that answers yes or no** to a transaction your app already built. State still exists, but it lives in the **datum** on a UTxO rather than inside the contract. **[Cardano for Ethereum developers](/docs/developers/cardano-for-ethereum-developers)** covers the rest. This track builds the model from scratch anyway.

Reviewer: ```suggestion
"Smart contract" means something different here. On Cardano, it is not a deployed program with storage that you call and that then acts. It is a **rule that answers yes or no** to a transaction your app has already built. State still exists, but it lives in the **datum** on a UTxO rather than inside the contract. **[Cardano for Ethereum developers](/docs/developers/cardano-for-ethereum-developers)** covers how Ethereum and Cardano development differ. This track teaches it from scratch.
```

:::

## What you'll be able to do

After this track you'll be able to:

- Split any Cardano app into two halves: the part your code prepares, and the part the chain enforces.
- Read a validator and say what it lets through and what it rejects.
- Choose what goes in a datum and what goes in a redeemer, and explain why.
- Write a validator, compile it, and get a script address out of the blueprint.
- Build the transactions that lock funds at that address and unlock them again, collateral included.
- Run a contract you wrote end to end on the test network, and watch it refuse a spend that breaks its rule.
- Put a deadline on funds, mint a token from a contract, and change data that's already on the chain.
- Publish a contract once instead of carrying it in every transaction, and let one contract read another's data without consuming it.
- Test a contract properly before it ever holds anything real.

## The lectures

> 1. **[On-chain vs off-chain](/docs/developers/onboarding/lectures/intermediate/on-chain-vs-off-chain)** — the two halves: your app prepares and the chain enforces.

Author: As we discussed in the meeting, we explain about the dapp architecture in the beginner, but I will bring to the on-chain vs off-chain page a quickly explanation about dapp and link to the beginner explanation.

2. **[Set up your tools](/docs/developers/onboarding/lectures/intermediate/tools)** — a language and compiler, and the contract project the next six lectures fill.
3. **[What a validator is](/docs/developers/onboarding/lectures/intermediate/what-is-a-validator)** — a yes/no gatekeeper that guards a locked UTxO.
4. **[Datum & redeemer](/docs/developers/onboarding/lectures/intermediate/datum-and-redeemer)** — the data you hand a contract.
5. **[The transaction context](/docs/developers/onboarding/lectures/intermediate/transaction-context)** — everything else a contract can look at before it decides.

> 6. **[Testing](/docs/developers/onboarding/lectures/intermediate/testing)** — tracing, unit tests and property-based tests, so the next two lectures can change the contract safely.

Author: Yes, got it. The text here meant "After your contract is tested, you can improve, change and the test guarantees that the rules defined before still work. But I will change as your suggestion.

7. **[Parameters](/docs/developers/onboarding/lectures/intermediate/parameters)** — a value built into the contract itself, before it has an address.
8. **[Validator purposes](/docs/developers/onboarding/lectures/intermediate/validator-purposes)** — spend and mint, under one script hash.

Those eight are the contract, and nothing after them changes it. The ninth is the other half, all at once:

9. **[Off-chain and frontend integration](/docs/developers/onboarding/lectures/intermediate/frontend-integration)** — derive the address, build every transaction, prove them offline, then connect a wallet and drive the vault from a page in the browser.

The last four are what you build with the machine:

10. **Handling time: vesting** — funds that can't move before a date, enforced without the contract ever reading a clock.
11. **Multi validators: a gift card** — one script guarding two different actions at once, and minting from a contract.
12. **Modifying state: an oracle** — changing data that's already on the chain.
13. **Reference inputs & reference scripts** — publish a contract once, and let one contract read another's data without consuming it.

> ## You write the vault, one lecture at a time

Reviewer: Also, why do lectures 1-9 have a special section? I'd either call it something generic like "The projects you'll build" or "How this onboarding flows" and explain more generically that we build 4 projects, that the first one takes 9 lectures to take it step-by-step, but that the rest are stand-alone lectures, etc.

Lectures 1 and 2 set up your workspace, `cardano-vault/`, with a half for each side, and leave you inside the contract project at `on-chain/vault/`. You stay there through lecture 8, so every Aiken command is the short kind: `aiken check`, `aiken add`, `aiken build`, with no paths to get wrong. Lecture 9 steps back up to the root, and that is the last folder change in the track. In between, each lecture explains one idea and has you add it to your contract: a validator that says yes to everybody, then the datum and redeemer, then the rule itself, then tests to hold it still, then a backup key, and finally a **mint** purpose so the vault can create its own token.

Each step is a few lines and one command, and each one ends with a clean `aiken check`. From **[testing](/docs/developers/onboarding/lectures/intermediate/testing)** onwards it also ends with a passing test suite, which is what makes the two lectures after it safe: both change a contract that already works.

> The four lectures after that work differently. Those contracts arrive finished, and the exercises have you break one and write the missing rule back.

Reviewer: There's a lot of information here that doesn't belong to the introduction and things to change:
- No need to specify the folder names; that's an implementation detail.
- No need to specify that "so every Aiken command is the short kind". We have to just do it, not talk about it.
- We keep talking about halves like: "with a half for each side", but we haven't yet explained the architecture of a dapp, so it doesn't make sense for a new user to read this.
- The "Each step is a few lines and..." paragraph is not the natural way we'd explain this. Besides, what does "which is what makes the two lectures after it safe" mean exactly? The previous lectures aren't safe? I know you tried to imply that with tests we check that the contract is safe(er), but it's not the same.
- "Those contracts arrive finished, and the exercises have you break one and write the missing rule back" is not natural, but more importantly, we're not providing the contracts finished; they have to write them themselves.

- **A provider key.** Your app now has to read UTxOs that are not in your wallet, the ones sitting at a contract's address, and work out what running a validator will cost before it sends anything. A wallet cannot do either, so you need a **[provider](/docs/developers/onboarding/lectures/beginner/providers-and-explorers)**. Get a free **[Blockfrost](https://blockfrost.io/)** Preview key.
- **A compiler for contracts.** You write the vault from lecture 3 onwards, so you need the toolchain for the language you pick:

<Tabs groupId="onchain">
<TabItem value="aiken" label="Aiken" default>

Install it from the **[Aiken installation guide](https://aiken-lang.org/installation-instructions)**. It takes about a minute.

</TabItem>
<TabItem value="scalus" label="Scalus">

A [Scalus](https://scalus.org/) version is coming soon. The idea is identical, only the language differs.

</TabItem>
</Tabs>

## The playground {#the-playground}

> Everything in these lectures is also finished and working in one example project, which we call the **playground**. It has every contract in the track, plus a small browser app that drives them: connect a wallet, mint and lock funds, unlock them again, put a deadline on funds, update an oracle.

Reviewer: ```suggestion
Everything in these lectures is also finished and working in one example project, which we call the **playground**. It has every contract in the track, plus a small browser app that drives them: connect a wallet, mint and lock funds, unlock them again, put a deadline on funds, update an oracle, etc.
```

You do not need it to follow the lectures. It is here for two reasons:

- **To see where you are going.** Run it once now, and the rest of the track is you rebuilding the first part of it yourself.

> - **To get unstuck.** Every exercise says the same thing at the end: the finished code is here.

Reviewer: ```suggestion
- **To get unstuck.** Every exercise solution is provided in the playground's code.
```

Download it, and start the app:

```bash
npx giget@latest gh:cardano-foundation/developer-portal/examples/onboarding/lectures/intermediate playground
```

<Tabs groupId="offchain">
<TabItem value="mesh" label="Mesh" default>

```bash
cd playground/vault/off-chain/mesh
npm install
cp .env.example .env      # then paste your Blockfrost Preview key into it
npm run dev
```

</TabItem>
<TabItem value="evolution" label="Evolution">

An [Evolution](https://github.com/IntersectMBO/evolution-sdk) version is coming soon. The idea is identical, only the library calls differ.

</TabItem>
</Tabs>

> Open the printed URL **in the browser where Lace is installed**. Connect, set up collateral, then **Lock 5 ADA** and **Unlock** it again. The **Mint & lock** button does the same thing but also creates a token under the contract's own policy, which is what **[validator purposes](/docs/developers/onboarding/lectures/intermediate/validator-purposes)** is about.

Reviewer: This concrete example should come after the explanation below about the folder structure and how it works.
